File: R/add_granularity_levels_up.R
#' Function add data to higher granularity levels
#'
#' This function generates values for higher granularity levels.
#' At the moment it takes A.ADM3 and adds them up to A.ADM2 and A.ADM1 and CH.
#'
#' @param df a data.frame where the values are included
#' @param list_ontologies a character vector with ontologies to be summed up to.
#'
#' @export
add_granularity_levels_up<-function(df,list_ontologies,list_dimensions=NULL){
  # TODO: This function needs period_value functionality later


  spatial_unit_table<-read.csv("data/spatialunits.csv")[,c(
    "spatialunit_ontology","spatialunit_hist_id","canton_hist_id","district_hist_id")]

  #put all the rows out that are not adm3
  df_out<-df
  #continue only with adm3 data
  df<-df[df$spatialunit_ontology=="A.ADM3",]

  df<-merge(df,spatial_unit_table,by=c("spatialunit_ontology","spatialunit_hist_id"),all.x=T,sort=F)

  if("A.ADM2" %in% list_ontologies){
    list_to_group<-c("district_hist_id","time_value",list_dimensions)
    df_adm2<-df %>% group_by(!!!syms(list_to_group)) %>% summarize(value=sum(value,na.rm=T))
    df_adm2$spatialunit_ontology<-"A.ADM2"
    df_adm2$spatialunit_hist_id<-df_adm2$district_hist_id
    df_adm2$spatialunit_name<-translate_to_spatial_unit_name(df_adm2,"de")
    df_adm2$period_value<-NA
    df_out<-rbind(df_out,df_adm2[,c(GLOBAL_TOTAL_LIST,list_dimensions)])
    df_adm2<-NULL
  }
  if("A.ADM1" %in% list_ontologies){
    list_to_group<-c("canton_hist_id","time_value",list_dimensions)
    df_adm1<-df %>% group_by(!!!syms(list_to_group)) %>% summarize(value=sum(value,na.rm=T))
    df_adm1$spatialunit_ontology<-"A.ADM1"
    df_adm1$spatialunit_hist_id<-df_adm1$canton_hist_id
    df_adm1$spatialunit_name<-translate_to_spatial_unit_name(df_adm1,"de")
    df_adm1$period_value<-NA
    df_out<-rbind(df_out,df_adm1[,c(GLOBAL_TOTAL_LIST,list_dimensions)])
    df_adm1<-NULL
  }
  if("CH" %in% list_ontologies){
    list_to_group<-c("time_value",list_dimensions)
    df_ch<-df %>% group_by(!!!syms(list_to_group)) %>% summarize(value=sum(value,na.rm=T))
    df_ch$spatialunit_ontology<-"CH"
    df_ch$spatialunit_hist_id<-0
    df_ch$spatialunit_name<-"Schweiz"
    df_ch$period_value<-NA
    df_out<-rbind(df_out,df_ch[,c(GLOBAL_TOTAL_LIST,list_dimensions)])
    df_ch<-NULL
  }

  return(df_out[,c(GLOBAL_TOTAL_LIST,list_dimensions)])
}

File: statbot_helpers/add_granularity_levels_up.py
import pandas as pd
import numpy as np
import os
import json
import sys 
import statbot_helpers as sbh

def add_granularity_levels_up(df,list_ontologies, path, list_dimensions = None):

    spatial_unit_table = pd.read_csv(os.path.join(path,"data/spatialunits.csv"), 
                                 usecols=["spatialunit_ontology","spatialunit_hist_id","canton_hist_id","district_hist_id"])

    spatial_unit_table[['canton_hist_id', 'district_hist_id']] = spatial_unit_table[['canton_hist_id', 'district_hist_id']].astype('Int64')


    df_out = df.copy()

    df = df[df['spatialunit_ontology'] == 'A.ADM3']

    df = pd.merge(df, spatial_unit_table, how='left', on=['spatialunit_ontology', 'spatialunit_hist_id'])

    with open(os.path.join(path,"constants/"+"constants.json")) as f:
        GLOBAL_TOTAL_LIST = json.load(f)

    allcols = list(GLOBAL_TOTAL_LIST['GLOBAL_TOTAL_LIST']) + list_dimensions
    
    if 'A.ADM2' in list_ontologies:
        list_to_group = ["district_hist_id","time_value"] + list_dimensions
        df_adm2 = df.groupby(by=list_to_group).agg({'value': ['sum']}).reset_index()
        df_adm2['spatialunit_ontology'] = 'A.ADM2'
        df_adm2['spatialunit_hist_id'] = df_adm2['district_hist_id']
        df_adm2['spatialunit_name'] = sbh.translate_to_spatial_unit_name(df_adm2, "de", os.path.join(path,"data/spatialunits.csv"))
        df_adm2['period_value'] = np.NAN
        df_adm2 = df_adm2[allcols]
        #df_out = df_out[allcols]
    
        df_adm2.columns = allcols
    
        df_out =  df_out.append(df_adm2, ignore_index=True)
        
    if 'A.ADM3' in list_ontologies:
        list_to_group = ["canton_hist_id","time_value"] + list_dimensions
        df_adm3 = df.groupby(by=list_to_group).agg({'value': ['sum']}).reset_index()
        df_adm3['spatialunit_ontology'] = 'A.ADM3'
        df_adm3['spatialunit_hist_id'] = df_adm3['cantont_hist_id']
        df_adm3['spatialunit_name'] = sbh.translate_to_spatial_unit_name(df_adm3, "de", os.path.join(path,"data/spatialunits.csv"))
        df_adm3['period_value'] = np.NAN
        df_adm3 = df_adm3[allcols]
        #df_out = df_out[allcols]
    
        df_adm3.columns = allcols
    
        df_out =  df_out.append(df_adm3, ignore_index=True)
        
    if 'CH' in list_ontologies:
        list_to_group = ["time_value"] + list_dimensions
        df_ch = df.groupby(by=list_to_group).agg({'value': ['sum']}).reset_index()
        df_ch['spatialunit_ontology'] = 'CH'
        df_ch['spatialunit_hist_id'] = 0
        df_ch['spatialunit_name'] = "Schweiz"
        df_ch['period_value'] = np.NAN
        df_ch = df_ch[allcols]
        #df_out = df_out[allcols]
    
        df_ch.columns = allcols
    
        df_out =  df_out.append(df_ch, ignore_index=True)
    
    return df_out